network/dag/state.go

```diff
@@ -0,0 +1,236 @@
+/*
+ * Nuts node
+ * Copyright (C) 2022 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package dag
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"time"
+
+	"github.com/nuts-foundation/nuts-node/core"
+	"github.com/nuts-foundation/nuts-node/crypto/hash"
+	"github.com/nuts-foundation/nuts-node/network/storage"
+	"github.com/nuts-foundation/nuts-node/vdr/types"
+	"go.etcd.io/bbolt"
+)
+
+const (
+	// boltDBFileMode holds the Unix file mode the created BBolt database files will have.
+	boltDBFileMode = 0600
+)
+
+// defaultBBoltOptions are given to bbolt, allows for package local adjustments during test
+var defaultBBoltOptions = bbolt.DefaultOptions
+
+// State has references to the DAG and the payload store.
+type state struct {
+	db           *bbolt.DB
+	graph        *bboltDAG
+	payloadStore PayloadStore
+	observers    []Observer
+	keyResolver  types.KeyResolver
+	publisher    Publisher
+	txVerifiers  []Verifier
+}
+
+// NewState returns a new State. The State is used as entry point, it's methods will start transactions and will notify observers from within those transactions.
+func NewState(dataDir string, verifiers ...Verifier) (State, error) {
+	dbFile := path.Join(dataDir, "network", "data.db")
+	if err := os.MkdirAll(filepath.Dir(dbFile), os.ModePerm); err != nil {
+		return nil, fmt.Errorf("unable to create BBolt database: %w", err)
+	}
+
+	// for tests we set NoSync to true, this option can only be set through code
+	var bboltErr error
+	db, bboltErr := bbolt.Open(dbFile, boltDBFileMode, defaultBBoltOptions)
+	if bboltErr != nil {
+		return nil, fmt.Errorf("unable to create BBolt database: %w", bboltErr)
+	}
+
+	graph := newBBoltDAG(db)
+
+	payloadStore := NewBBoltPayloadStore(db)
+	newState := &state{
+		db:           db,
+		graph:        graph,
+		payloadStore: payloadStore,
+		txVerifiers:  verifiers,
+	}
+
+	publisher := NewReplayingDAGPublisher(payloadStore, graph).(*replayingDAGPublisher)
+	newState.RegisterObserver(func(ctx context.Context, transaction Transaction, payload []byte) {
+		publisher.publishMux.Lock()
```

Comment:
publisher and state are a bit too tangled here: `state` calls functions on `publisher`, but knows it has to use some certain mutex of it to make them safe. I'd say publisher has to register those observers then, or let `transactionAdded` and `payloadWritten` lock themselves, as required?